### ISSUE TITLE
Change Lock entity to Unavailable when MQTT is disconnected 

### ISSUE THREAD

**Issue by leranp** [technyon/nuki_hub#227]
Hi,
Love this component, but there is something that i think that needs to be changed, when the MQTT is disconnected for some reason, the lock entity still shows status that can be worng.
In all of my MQTT base devices, when the MQTT is not connected, the device goes to Unavailable State

**Comment by technyon** [comment]
It wouldn't make sense to change the lock state if MQTT is disconnected. The lock doesn't necessarily change state just because MQTT is disconnected.

What you want to monitor is the "maintenance/mqttConnectionState" node, which reflects wether or not the ESP is connected ... which reminds me I'll have to add this to the readme.

**Comment by mundschenk-at** [comment]
I disagree, in HA that absolutely would make sense. When the ESP is disconnected from MQTT, the lock state is `unavailable`. Adding it to auto-discovery should be pretty straightforward, I can try to come up with something.

**Comment by technyon** [comment]
ok, understood. However @leranp didn't mention this is HA related.

**Comment by leranp** [comment]
Yes, i am talking about HA

**Comment by technyon** [comment]
@mundschenk-at Let's update the auto discovery then. We can release this together with the fix for the auto discovery warnings.

**Comment by mundschenk-at** [comment]
Sorry for the delay, I'll do the write-up this weekend.

**Comment by technyon** [comment]
@mundschenk-at ok

**Comment by mundschenk-at** [comment]
So in YAML you'd need to add this to auto-discovery:
```yaml
      availability:
        - topic: <prefix>/maintenance/mqttConnectionState
```

No need to set the payloads, as they default to `online` and `offline` anyway. So for the JSON payload, it would be:

```json
"avty": [{"t": "~/maintenance/mqttConnectionState"}]
```

I am using the `availability` syntax here because it would allow us to add several availability topics, e.g. not only for the MQTT connection but also for the BLE connection - the state of the lock is only accurate if there was recent-ish BLE connection and MQTT is connected. You could set such a topic to `offline` when a BLE restart is necessary. That would make such interruptions visible in the HA logbook.

**Comment by mundschenk-at** [comment]
Anyway, it was just a thought, probably too niche a use-case. The MQTT availability topic should probably be added to all auto-discovery entities except the `MQTT connected state`.

**Comment by technyon** [comment]
I've added availability to the json as suggested. Please check and confirm that it works. So far only for the lock, I'd added it to the other entities if it's confirmed it works for the lock.

[nuki_hub-8.27-pre-2.zip](https://github.com/technyon/nuki_hub/files/12710960/nuki_hub-8.27-pre-2.zip)

**Comment by mundschenk-at** [comment]
I tried it, works fine for the lock, but It breaks the opener because the connection state is not available under the opener prefix. Not sure if you can separate them with inheritance? 

I do have a suggestion how to handle this (with my long-promised write-up of a better auto-discovery device split), but that probably needs some discussion.

**Comment by mundschenk-at** [comment]
This would be the quick fix, but it is kind of dirty. Or you could simply reference the `mqttConnectionState` under the lock prefix.

IMHO nicer would be to move all the Nuki Hub/ESP-specific topics to their own HA "hub" device (and maybe MQTT topic, but that would be a breaking change of course) and have the Lock and Opener be connected devices to that "hub". (I'd have to look into it whether we would then still need the separate availability topic in the Lock and Opener device definitions, maybe we do.)

**Comment by technyon** [comment]
All paths are relative right? Should be easy enough to point to the lock maintenance topic.

**Comment by technyon** [comment]
Please check if it works like this.

[nuki_hub-8.27-pre-3.zip](https://github.com/technyon/nuki_hub/files/12846294/nuki_hub-8.27-pre-3.zip)

**Comment by mcervar** [comment]
Hi 

I just tried it, I dont see error in HA logs anymore, thx!

**Comment by leranp** [comment]
> Please check if it works like this.
> 
> [nuki_hub-8.27-pre-3.zip](https://github.com/technyon/nuki_hub/files/12846294/nuki_hub-8.27-pre-3.zip)
> 
> 

Looks fine for me

**Comment by technyon** [comment]
@mundschenk-at Does it fix the issue with the opener?

**Comment by mundschenk-at** [comment]
> @mundschenk-at Does it fix the issue with the opener?

Unfortunately, no. I missed your question about paths being relative, to which the answer would be no (or rather: MQTT topics are not paths that can be navigated). So 
```json
"availability": {
  "topic": "nukiopener/../nuki/maintenance/mqttConnectionState"
}
```
is incorrect, it would need to be
```json
"availability": {
  "topic": "nuki/maintenance/mqttConnectionState"
}
```

**Comment by mundschenk-at** [comment]
Sorry I wasn't more explicit yesterday, the new version works fine.

**Comment by technyon** [comment]
Check release 8.27

**Comment by bcutter** [comment]
OK I updated to 8.27 some days ago (same day it became available). Now that I restarted HA for the first time since then, all my locks are shown as `unavailable`, even they are reachable and at the same time the MQTT connected entities show `on`.

Not sure what screwed things up - I've NEVER ever seen the lock entities in `unavailable` state before the 8.27 update.
Will try to downgrade one Nuki hub to 8.26 for now and see if that changes things.

![grafik](https://github.com/technyon/nuki_hub/assets/13799156/af495d66-460d-4d89-b02f-8c0555046120)

![grafik](https://github.com/technyon/nuki_hub/assets/13799156/3f994b4d-ee68-4753-a8b0-435d3f99d02f)

![grafik](https://github.com/technyon/nuki_hub/assets/13799156/aa29ce9a-a584-41ad-bcb5-4a95c0d67789)

**Comment by technyon** [comment]
What path did you configure for the lock? And could you share the content of the HA discovery topic for the lock?

**Comment by bcutter** [comment]
Could you please guide me precisely on how to get that information?

Afaik I never changed/customized things so everything _should_ be default.

**Comment by technyon** [comment]
As a start, maybe post the content of the system information page.

Which MQTT broker are you using?

**Comment by mundschenk-at** [comment]
> Could you please guide me precisely on how to get that information?

In HA: Settings > Devices & Services > MQTT > Your lock device > 3 dots > Download diagnostics

**Comment by bcutter** [comment]
System information page (masked)
```
NUKI Hub version: 8.26
run: true
deviceId: XXXXXXXXXXX
deviceIdOp: XXXXXXXXXXX
mqttbroker: xxx.xxx.xxx.xxx
mqttport: 1883
mqttuser: ***
mqttpass: ***
mqttlog: false
lockena: true
mqttpath: nuki_doorone
openerena: false
mqttoppath: 
maxkpad: 5
opmaxkpad: 
mqttca: 
mqttcrt: 
mqttkey: 
hassdiscovery: homeassistant
dhcpena: true
ipaddr: xxx.xxx.xxx.xxx
ipsub: 255.255.255.0
ipgtw: xxx.xxx.xxx.xxx
dnssrv: xxx.xxx.xxx.xxx
nwhw: 1
rssipb: -1
hostname: ESP-Nuki-Hub-DoorOne
nettmout: -1
restdisc: true
rstbcn: 60
lockStInterval: 3600
configInterval: 3600
batInterval: 7200
kpInterval: 1800
kpEnabled: true
accLvl: 
regAsApp: false
nrRetry: 5
rtryDelay: 100
crdusr: ***
crdpass: ***
pubauth: true
pubdbg: false
prdtimeout: 60
hasmac: false
macb0: 
macb1: 
macb2: 
MQTT connected: Yes
Lock firmware version: 3.7.7
Lock hardware version: 5.11
Lock paired: Yes
Lock PIN set: Yes
Lock has door sensor: No
Lock has keypad: Yes
Network device: Built-in Wifi
Uptime: 1088 minutes
Heap: 85424
Stack watermarks: nw: 6104, nuki: 360, pd: 248
Restart reason FW: RestartOnDisconnectWatchdog
Restart reason ESP: ESP_RST_SW: Software reset via esp_restart.
```

As the MQTT -> lock -> burger menu -> diagnostics output is very very chatty, here's what I think you might have asked for:
```
        {
          "entity_id": "lock.schloss_doorone",
          "subscriptions": [
            {
              "topic": "nuki_doorone/lock/binaryState",
              "messages": [
               ...................................
                {
                  "payload": "locked",
                  "qos": 0,
                  "retain": 0,
                  "time": "2023-10-24T06:18:06.074445+00:00",
                  "topic": "nuki_doorone/lock/binaryState"
                },
                {
                  "payload": "unlocked",
                  "qos": 0,
                  "retain": 0,
                  "time": "2023-10-24T06:29:16.264524+00:00",
                  "topic": "nuki_doorone/lock/binaryState"
                }
              ]
            }
          ],
```

Hope that helps.

**Comment by mundschenk-at** [comment]
No, the discovery topic should be something like `homeassistant/lock/<lockid>/<entityname>/config` (under the key `discovery_data`). However, I've just checked mine and there is some corruption going on. The lock entity shows an incorrect discover topic (`~/maintenance/mqttConnectionState`). However, this looks like a display issue with the HA MQTT integration. The actual discovery topic still exists on my broker.

**Comment by bcutter** [comment]
One picture maybe tells more than 1.000 words. Here's once again the current status:

![grafik](https://github.com/technyon/nuki_hub/assets/13799156/43325e95-afec-4558-a879-04142229b87e)

**Comment by technyon** [comment]
Could you post the whole content of that topic?

**Comment by bcutter** [comment]
I die so already for one of the two locks, see at the bottom of this post https://github.com/technyon/nuki_hub/issues/227#issuecomment-1777984588

**Comment by bcutter** [comment]
Yes everything I provided is from Nuki Hub 8.26. I don't see "avty". Where exactly should I look at?

Yes, this is as expected:
![grafik](https://github.com/technyon/nuki_hub/assets/13799156/b1b73df6-1fdc-4cf7-834b-150cdd45e982)

Existing and updating regularly. On 8.26 (⚠️).

Not sure what else I could provide...

**Comment by technyon** [comment]
That's what has been added with 8.27, so you won't see it with 8.26

**Comment by bcutter** [comment]
OK. So I guess I can't add something helpful, do I?

**Comment by technyon** [comment]
Well... you could update to 8.27 and copy/paste what's there, then go back to 8.26

**Comment by bcutter** [comment]
Updated one Nuki Hub again to 8.27. autodiscovery topic (`config`)...
![grafik](https://github.com/technyon/nuki_hub/assets/13799156/b5632f7e-e603-4f11-85d4-17a694c5c951)
...now has:
![grafik](https://github.com/technyon/nuki_hub/assets/13799156/641cb91b-1f40-4d3a-b7b9-fbc19c7f8f55)

And that one still exists of course:
![grafik](https://github.com/technyon/nuki_hub/assets/13799156/91fa7c6e-e529-4585-9e06-78acdcc1e113)

So I assume the `avty` is not available or not correct or something else during HA restart. Otherwise I can't imagine why the lock entity renders unavailable (which for the records it did not during the short update 8.26-8.27 as I did not restart HA).

Let's compare auto discovery topic:
8.26:
```
{"dev":{"ids":["nuki_XXXXXXXX"],"mf":"Nuki","mdl":"SmartLock"},"~":"nuki_doortwo","name":"Door Two","unique_id":"XXXXXXXX_lock","cmd_t":"~/lock/action","pl_lock":"lock","pl_unlk":"unlock","pl_open":"unlatch","stat_t":"~/lock/binaryState","stat_locked":"locked","stat_unlocked":"unlocked","opt":"false"}
```
8.27:
```
{"dev":{"ids":["nuki_XXXXXXXX"],"mf":"Nuki","mdl":"SmartLock","name":"Door Two"},"~":"nuki_doortwo","name":null,"unique_id":"XXXXXXXX_lock","cmd_t":"~/lock/action","avty":{"t":"~/maintenance/mqttConnectionState"},"pl_lock":"lock","pl_unlk":"unlock","pl_open":"unlatch","stat_t":"~/lock/binaryState","stat_locked":"locked","stat_unlocked":"unlocked","opt":"false"}
```

Comparing both I only discovered that
a) `name` exists two times now with 8.27,
b) first `name` has another position but is correct and
c) second `name` is on the same place as with 8.26 but is `null` (is that suspicious or relevant at all?)

**Comment by mundschenk-at** [comment]
That's on purpose due to how recent HA versions handle device vs. entity names.

**Comment by bcutter** [comment]
OK. So could that be a definite issue, running an older HA version?

**Comment by bcutter** [comment]
> Normally `<prefix>/maintenance/mqttConnectionState` will only change to `offline` if the hub is disconnected from the MQTT broker (by pulling power for example). A reboot of the ESP will not change the topic to `offline` (probably because it is too fast for the session timeout).

Yes, but as initially described here https://github.com/technyon/nuki_hub/issues/227#issuecomment-1773373364 with

> all my locks are shown as `unavailable`, even they are reachable and at the same time the MQTT connected entities show `on`.

**that's not the issue. MQTT connection state completely fine, but lock entities `unavailable`.**

> Which version are you running?

Im running HA Core 2023.3.6 at the moment.

So now it's up to you to figure out
a) why this happens and
b) if/why only for me (what's the update pace, how often has 8.27 been downloaded, how many of those use HA etc.)

I don't see what else I could provide. 8.26 everything's running fine, 8.27 screws up lock entities once HA restarts. No idea what you changed in detail in the 8.27 code.

**Comment by technyon** [comment]
From the issues received here, I'd say the majority of users use HA. The code that changed in details doesn't matter all that much, what's important is what's in MQTT. "avty" was added so HA knows when the ESP goes offline. Why it doesn't work in your case is somewhat of a mystery, since everything looks fine.

We can try to figure out what's different in your configuration. What MQTT broker are you using, maybe a specific broker causes problems. Also you have an underscore in the path ... that shouldn't matter but then it could. Maybe as a test change the path to not use an underscore.

**Comment by bcutter** [comment]
I'm just using the default Mosquitto broker shipped as addon for HA (https://github.com/home-assistant/addons/tree/master/mosquitto).

Well, instead of ruining my productive system and making my other houshold members sharpen their knifes, I hope you experts have the possibility of a test system. So you could just add a (very very VERY likely unproblematic `_`) to one of your locks and give it a go.

**Comment by mundschenk-at** [comment]
Well, you might also want to update your HA version to something more recent. That should not be the issue, but as is, there are a lot of variables here. You can't expect people here to reproduce your exact system.

**Comment by bcutter** [comment]
> Well, you might also want to update your HA version to something more recent. That should not be the issue, but as is, there are a lot of variables here.

Yes, I will update HA once there's the needed time. I also don't think this will make a difference as I checked all breaking changes in all monthly release notes between 2023.3 and 2023.10 - nothing in MQTT which could have an impact in this way it renders (lock) entities unavailable as false positive.

> You can't expect people here to reproduce your exact system.

No of course not. But one can the pros have more test possibilities and rule out low-level assumptions like a certain character in a name :-)

**Comment by bcutter** [comment]
Small note: found something in HA 2023.10 release notes which sounds kinda interesting:

https://www.home-assistant.io/blog/2023/10/04/release-202310/#breaking-changes

> An MQTT lock with a configured state topic will initialize with state unknown instead of state unlocked unless the lock is set to optimistic mode.

> MQTT locks set to optimistic mode will still be initialized with the unlocked state. You should check if your automations are affected.

While I'm not 100 % sure what the detailed impact of this change on lock entities in HA is, it's impossible my issue is caused by this - as I'm still few release cycles behind that 2023.10 update. Maybe you want do take a look at this change anyway in case you would need to configure something.

**Comment by technyon** [comment]
I'm not really sure how to reproduce this. Maybe updating your system first would make sense so we can eliminate at leasta one variable. Other than that, we could think about a way to disable publishing avty.

**Comment by bcutter** [comment]
Going through the updates, but it takes some time, probably weeks.
2023.4 made no difference so far (and tbh I doubt it will be fixed with a HA update).

So annoying when updates break (core) things... 8.26 was running so smooth and stable. I get the "want have" factor for having lock entities unavailable depending on the MQTT connection state, but well... not working.

**Maybe we can have this avty thing as _config option_?** So being on as default this way everyone could simply decide on his/her own.

**Comment by mundschenk-at** [comment]
Well, if you mean @alexdelprete and me, yes, we are using EMQX. Not sure about everyone else. The broker should not make any difference here though, unless it is somehow possible to disable LWT in Mosquitto. 

@bcutter Please look at your HA log to see if there is any issue with MQTT discovery messages not being parsed - the component is  quite chatty. If not, try to enable debug logging for the `mqtt` component.

**Comment by alexdelprete** [comment]
> I wonder if it could be somehow related to the broker. I think everyone else uses EMQX ... is this correct?

It's not broker related. It's an issue with @bcutter HA not parsing tha `availability` topic correctly for some unknown reason. Also, I would use `availability_topic` vs `availability`, since we only use one topic and not multiple topics to determine the availability status. 

```json
"avty_t": "~/maintenance/mqttConnectionState"
```

@bcutter can you try an older version of NH, pre 8.27, to exclude that the `availability` topic (introduced in 8.27) is an issue in your case?

**Comment by alexdelprete** [comment]
> While I'm not 100 % sure what the detailed impact of this change on lock entities in HA is

That breaking change simply meanst that on HA startup, MQTT Locks have an unknown state, until the state is correctly determined by the relevant MQTT topic. Basically HA doesn't assume a default state, unless optimistic mode is configured.

**Comment by bcutter** [comment]
Yes indeed, 8.26 works without any issues - because it does not have the avty set yet.

So many different approaches here (plenty of activity, appreciate that), not sure what to do / check next on my side. If you experts isolated the core issue please let me know what I can contribute.

I'm kinda sick and tired of updating NH to 8.27 and restarting HA just to see all my lock entities fail (unavailable) again, I've gone through that process several times lately. If I really need to do that one more time I want to know exactly what to check/do/provide you here.

**Comment by alexdelprete** [comment]
> If you experts isolated the core issue please let me know what I can contribute.

The core issue seems to be specific to your setup, I haven't seen other reports after 8.27 with your same issue. Please correct me if I missed any. My installation works perfectly fine with 8.27 and my lock doesn't go unavailable.

> If I really need to do that one more time I want to know exactly what to check/do/provide you here.

I suggested to change `avty` to `avty_t`, so when dev provides that test version you might want to test it. What version of HA are you running now? Can you also describe where HA and MQTT broker are running? RPi, VM, LXC, etc. Thanks.

**Comment by alexdelprete** [comment]
@bcutter can you tell me if the `mqttConnectionState` topic has the retain flag in the broker? I'm using MQTT Explorer to navigate the topics in the broker here.

![image](https://github.com/technyon/nuki_hub/assets/7027842/4470864b-1e67-4942-87a9-7153dff70dca)

**Comment by technyon** [comment]
> I suggested to change `avty` to `avty_t`, so when dev provides that test version you might want to test it. What version of HA are you running now? Can you also describe where HA and MQTT broker are running? RPi, VM, LXC, etc. Thanks.

I think you can test it without a new binary. Just start NUKI Hub and let it publish the discovery topic, then manually change avty to avty_t using some MQTT client. NUKI Hub will only overwrite the discovery topics on boot.

**Comment by mundschenk-at** [comment]
OK, after sleeping over it, I now know why @bcutter's lock entities are unavailable with 8.27 - it's got nothing whatsoever to do with the availability topic, but simply a the same as https://github.com/technyon/nuki_hub/issues/246: The new empty `name` in the auto-discovery topic is invalid before Home Assistant 2023.8.0.

**Comment by alexdelprete** [comment]
> The new empty `name` in the auto-discovery topic is invalid before Home Assistant 2023.8.0.

So PR #225 causes the entities to be unavailable in HA < 2023.8.0 ?

If confirmed, we should put a note in readme of the repo, specifying HA 2023.8.0 as the minimum supported version. Maybe a pre-requirements paragraph.

Also, when these breaking changes are implemented, they should be highlighted as breaking in release notes so that users know they need to test it first.

**Comment by bcutter** [comment]
Great finding, congratulations 👏🏼 👍🏼

Why is it empty at all? Why not setting it?

That way it would resolve any dependency of a HA version, no breaking change notice needed (which is now too late anyway for 8.27) etc.

**Comment by alexdelprete** [comment]
> Why is it empty at all? Why not setting it?

It's an HA breaking change: read #217 

![image](https://github.com/technyon/nuki_hub/assets/7027842/073c751f-67e1-43da-bb05-bb60f49b3640)

> That way it would resolve any dependency of a HA version, no breaking change notice needed

That's not true, if HA changes something components/integrations need to adapt to those changes, hence minimum supported versions of HA need to be specified.

> which is now too late anyway for 8.27

not true either, there might be users with old HA versions that will install 8.27 and will read the instructions, so minimum requirements should always be mentioned since there are dependencies.

**Comment by mundschenk-at** [comment]
@alexdelprete: Right, this should in the README but we didn't really know until now. @technyon doesn't use HA at all and I'm always on the latest version (since I find quickly adapting my configuration to any breaking changes means much fewer issues in the long run). Unfortunately, the Powers That Be at HA didn't highlight the fact that this auto-discovery change would prevent it from working with older versions of HA. 

As a workaround, we could add `name: Lock` for the `lock` entities instead, which would result in the entity name  `lock.your_device_lock` and a friendly name of `Your Device Lock`, but personally I think a simple note in the README should suffice at this point.

@bcutter:
> Why is it empty at all? Why not setting it?
> 
> That way it would resolve any dependency of a HA version, no breaking change notice needed (which is now too late anyway for 8.27) etc.

Because that's the way it is intended: https://community.home-assistant.io/t/psa-mqtt-name-changes-in-2023-8/

If you want to stay on something older than 2023.8.0, you can always configure the `lock` entity yourself using YAML (even if you use auto-discovery for the other entities).

**Comment by alexdelprete** [comment]
> Right, this should in the README but we didn't really know until now. @technyon doesn't use HA at all and I'm always on the latest version

we adopted a change due to HA recommendations, so we should mention that as a breaking change. we knew it when we implemented it. We were relying on the new 2023.8.0 way of working, to avoid warnings. Doesn't matter if @technyon uses HA or not, we knew the impact when we adopted those changes.

The general readme should always contain pre-requirements, and in the release notes breaking changes (such as a pre-requirement change) should be highlighted properly.

It's a general advice for the future. :)

**Comment by mundschenk-at** [comment]
Well, at oeast I didn't know that the auto-discovery message would simply fail to parse on HA < 2023.8.0 (instead of using some sane default) 🤷

**Comment by alexdelprete** [comment]
> at least I didn't know that the auto-discovery message would simply fail to parse on HA < 2023.8.0

I understand. But in general, if you adopt a change based on HA version, you have to mention that as an important change and mention it. It's no big deal...but in the future we should not rely on intuitions or memory, we'd waste a lot of time for these issues. :)

**Comment by bcutter** [comment]
So to sum up for my current stuck-on-NH-8.26-situation: I need to upgrade to HA Core 2023.8 to be able using NH 8.27. There will be no changes made on NH side except leaving a breaking change note in the release notes description.

Correct?

Not the best outcome (as I won't be able to upgrade HA Core to 2023.8 before mid of December at best) for me but at least something I can count with.

**Comment by mundschenk-at** [comment]
You could also upgrade to 8.27 and configure the `lock` entity using YAML instead of relying on auto-discovery. But yes, staying on 8.26 until you can upgrade to HA >= 2023.8.x is probably the more easy route. It's not that you are missing out on much.

**Comment by bcutter** [comment]
I think I need to / will manage to look at this in my custom admin update dashboard for a while, sitting next to HA Core.

![grafik](https://github.com/technyon/nuki_hub/assets/13799156/a0e28340-09b8-4647-ba3a-1b7f20f6922d)
![grafik](https://github.com/technyon/nuki_hub/assets/13799156/5993ed03-16c2-484d-aaba-dde3a003c5b8)

**Really glad you found the root for this** and I don't need to spend more time on testing etc. 👍

**Comment by technyon** [comment]
Hi. Thanks for figuring this out, I'll add a note to the readme.

**Comment by alexdelprete** [comment]
> I need to upgrade to HA Core 2023.8 to be able using NH 8.27. There will be no changes made on NH side except leaving a breaking change note in the release notes description.

Why should NH make a change? It's in line with HA, your installation is not. 8.27 was made specifically for users who upgraded to latest HA versions and had that warning, and that's not you. So you should remain with version < 8.27.

It's not always possible to be background compatible when HA changes something an integration uses in core.

**Comment by alexdelprete** [comment]
> Hi. Thanks for figuring this out, I'll add a note to the readme.

Jan, could you also put a note in the release notes of 8.27? For users who will install it later...

**Comment by bcutter** [comment]
> It's in line with HA, your installation is not.

Well, expecting everyone uses latest versions of every component in the individual smart home eco system is a heavily dangerous state of mind imo.

Ideally software can technically avoid to install or even show updates when requirements are not met. Like in HACS where maintainers can precisely specify the requirements. Of course this is not possible for NH as it's a standalone project/firmware which makes it much more difficult to be as compatible as possible to the relevant smart home systems making use of the MQTT source. (sidenote: which is one reason for some wishing NH would be part of ESPHome so all these problems are semi-automatically taken care of, plus much smarter update management etc.)

So +1 for at least leaving notes in readme and release notes. I read them always, sometimes twice before upgrading.

Nothing more to add from my simple user / non-dev perspective. Thanks again for discovering this and taking care of.

Basically many of these posts here (starting with my first likely) belong to #217, not #227 as we know now.

**Comment by alexdelprete** [comment]
> Well, expecting everyone uses latest versions of every component in the individual smart home eco system is a heavily dangerous state of mind imo.

We don't expect anything, we integrate with HA development cycle: if HA makes a breaking change, ALL integrations (not only NH) impacted will need to adapt.

It's pretty clear you are not a dev so you don't understand the lifecycle and the interdependencies of a component that relies on another software, and it's a very dangerous state of mind having such a strong opinion about something you clearly have no competencies or comprehension about.

You, as a user, can decide to stay on HA 0.19 of some years ago, but if you decide to do so, don't expect component developers to make sure your installation works.

Furthermore, there's no reason for you to upgrade to 8.27, since 8.26 works perfectly well for you, with no HA warnings. That is not the case for users on latest HA versions. So you're not impacted in any way functionality wise.

Our mistake was not clearly stating in docs and release notes about the new pre-requirements on HA version.

> So +1 for at least leaving notes in readme and release notes. I read them always, sometimes twice before upgrading.

That's a very good practice, kudos to you. Unfortunately not many users do actually read them. They simply upgrade and expect everything to work, then start bragging about things that changed.

Thanks for your feedbacks, and next time before interpreting incorrectly why devs make some decisions and judging their state of minds, ask for more information, so you will understand the context and you will avoid potentially offending people because you ignore important facts.

Cheers.

**Comment by technyon** [comment]
I've added a sentence or two to the readme and the release notes. I think there's little choice but to adopt these changes - although I have to say I'm not sure if it was really necessary to make this a breaking change from HA side. They could've just ignored the name and recommend not to put it for new integrations.

**Comment by alexdelprete** [comment]
> although I have to say I'm not sure if it was really necessary to make this a breaking change from HA side. They could've just ignored the name and recommend not to put it for new integrations.

I agree. But HA devs are known to make these kind of decisions ignoring all the consequences to existing integrations / automations /etc. If you read the HA community, you'll find a lot of complaints about this. Unfortunately many integration/component devs take the blame for these decisions, while the root cause is bad decisions in HA.

But that's the way they see things, not much to do other than adapt, if you want to use HA, that I still believe is the best platform (for me obviously). :)

**Comment by technyon** [comment]
Well, they could act wiser, but it is as it is. I'll close this issue for now, since all problems are resolved (or at least understood).

**Comment by bcutter** [comment]
> Hi. Thanks for figuring this out, I'll add a note to the readme.

@technyon: General question: are any requirements in regards of a specific minimum Home Assistant version listed in every Nuki Hub release?

Like you did back then with the [8.27 release](https://github.com/technyon/nuki_hub/releases/tag/8.27)

> Note: This release needs at least Home Assistant version 2023.8.0 to function properly

**Comment by iranl** [comment]
This was added specifically because of changes to HA auto discovery that made Nuki Hub incompatible with HA version before 2023.8. Every HA version since is supported, although Nuki Hub does make use of recently added upgrades to the MQTT lock sensor. I don't see a reason to mention compatibility on every release unless the minimum version changes.

**Comment by bcutter** [comment]
OK, so the 8.27 release was the only one in need of a minimum HA version. That's the only thing I needed to know. Because of my old HA version I was stuck at NH v8.26. Now finally running HA 2023.8 (I'm keeping up slowly... ;-) ) I was wondering if it's now only "safe" to update to 8.27 or even up directly to the latest 8.35 release, without any further surprises.

**Comment by bcutter** [comment]
> although Nuki Hub does make use of recently added upgrades to the MQTT lock sensor

@iranl which are those? Do I need a "more recent" HA version to take advantage of those?

Just asking because still racking my brain because of https://github.com/technyon/nuki_hub/issues/434

**Comment by iranl** [comment]
Home Assistant 2024.6 

See https://www.home-assistant.io/blog/2024/06/05/release-20246/

> MQTT Lock now supports the lock open and opening states. When a lock is used in optimistic mode, its state will become open instead of unlocked when it is opened.
